### PR TITLE
Implementation of generate_data_key KMS Endpoints

### PR DESCRIPTION
**Pull request by jblacker**
This pull request implements the `generate_data_key` & `generate_data_key_without_plaintext` endpoints for the KMS service, along with tests.

I also refactored the "encryption"/"decryption" components so they can be reused more easily by other endpoints within the service.  I don't typically program in Python, so all comments/criticism are welcome!

Additionally, I've updated the KMS section of the implementation overage document.

I created this PR in response to my ticket #1518

### CODE REVIEW

File: moto/kms/responses.py [hunk 1]
```diff
@@ -220,19 +220,50 @@ def list_key_policies(self):
...
     def encrypt(self):
-        """
-        We perform no encryption, we just encode the value as base64 and then
-        decode it in decrypt().
-        """
         value = self.parameters.get("Plaintext")
         if isinstance(value, six.text_type):
             value = value.encode('utf-8')
-        return json.dumps({"CiphertextBlob": base64.b64encode(value).decode("utf-8")})
+        return json.dumps({"CiphertextBlob": self.kms_backend.encrypt(value)})
```
Comment (JackDanger) [inline moto/kms/responses.py]: Nice refactor - thanks for fixing this!

**Comment by JackDanger** [comment]
There is at least one test failure but the change otherwise looks good! Once the tests pass let's merge this.

**Comment by jblacker** [comment]
@JackDanger Thanks.  I saw the failure.  The length of the generated key isn't as long as expected.  I'll get that fixed up later today/tomorrow.

**Comment by mattsb42-aws** [comment]
@jblacker @JackDanger What is left that needs to be done to get this merged? Anything I can do to get it moving again?

I have a variety of improvements I would like to add to the KMS mocks, but would like to have the changes in this PR present beforehand.

**Comment by mattsb42-aws** [comment]
Actually, looking over the diffs, it looks like the model changes made their way in to `master` through other means. I can easily add in the generate data key methods independently if needed.

**Comment by dargueta** [comment]
Are there plans to get this merged in? I'm running into problems hacking my way around it.

**Comment by mattsb42-aws** [comment]
I think this PR can be closed. This surface area was addressed in #2397. IIRC, the changes between when this PR was originally cut and when I did #2397 were sufficient that it was simpler to start anew. I also was able to make the whole simulation much simpler by just actually doing the crypto. :)

**Comment by mattsb42-aws** [comment]
> I'm running into problems hacking my way around it.

Are you using the latest release? IIRC the changes from #2397 were released in 1.3.14.

**Comment by dargueta** [comment]
Breaks running Python 3.7 on both OSX and Ubuntu 18.04 (Bionic) with the following dependencies

```
boto3==1.12.4
botocore==1.15.4
moto==1.3.14
```

Getting 403 a lot but it's pretty flaky

**Comment by mattsb42-aws** [comment]
That...is odd. Could you open an issue with a replication case and tag me?

**Comment by spulec** [comment]
Thanks; closing per comments from @mattsb42-aws

**Comment by dargueta** [comment]
@mattsb42-aws Of course I can't repro it now. o__O

**Pull request closed without merging**